errai-navigation/src/main/java/org/jboss/errai/ui/nav/client/local/Navigation.java

```diff
@@ -444,49 +448,71 @@ public void onAttachOrDetach(final AttachEvent event) {
 
+        accessEvent.fireAsync(component, new LifecycleCallback() {
           @Override
           public void callback(final boolean success) {
             if (success) {
-              locked = true;
+              NavigationControl showControl = new NavigationControl(Navigation.this, new Runnable() {
+                @Override
+                public void run() {
+                  try {
+                    // Fire IOC lifecycle event to indicate that the state of the
+                    // bean has changed.
+                    // TODO make this smarter and only fire state change event when
+                    // fields actually changed.
+                    stateChangeEvent.fireAsync(component);
+
+                    setCurrentPage(request.pageNode);
+                    currentWidget = componentWidget;
+                    currentComponent = component;
+                    navigatingContainer.setWidget(componentWidget);
+                    request.pageNode.pageShown(component, request.state);
+                  } finally {
+                    locked = false;
+                  }
+
+                  handleQueuedRequests(request, fireEvent);
+                }
+              }, new Runnable() {
+                @Override
+                public void run() {
+                  locked = false;
+                }
+              });
+
               try {
+                locked = true;
                 hideCurrentPage();
-                request.pageNode.pageShowing(component, request.state);
-
-                // Fire IOC lifecycle event to indicate that the state of the
-                // bean has changed.
-                // TODO make this smarter and only fire state change event when
-                // fields actually changed.
-                stateChangeEvent.fireAsync(component);
-
-                setCurrentPage(request.pageNode);
-                currentWidget = componentWidget;
-                currentComponent = component;
-                navigatingContainer.setWidget(componentWidget);
-                request.pageNode.pageShown(component, request.state);
-              } finally {
+                request.pageNode.pageShowing(component, request.state, showControl);
+              }
+              catch (Exception ex) {
                 locked = false;
+                throw ex;
               }
-
-              handleQueuedRequests(request, fireEvent);
-            }
-            else {
+            } else {
               request.pageNode.destroy(component);
             }
           }
         });
       }
     });
 
-    if (currentPage != null && currentWidget != null && currentComponent != null && currentWidget.asWidget() == navigatingContainer.getWidget()) {
-      currentPage.pageHiding(Factory.maybeUnwrapProxy(currentComponent), control);
-    }
-    else {
+    if (!hiding && currentPage != null && currentWidget != null && currentComponent != null && currentWidget.asWidget() == navigatingContainer.getWidget()) {
```

Comment:
Ah, I think I understand now. Without this `hiding` boolean, calling redirect in a `@PageHiding` method causes a full navigation, meaning another call to `@PageHiding`.

Here's what I recommend: get rid of `hiding` and pass in an `interrupt` Runnable to the contstructor of `control` with a run method as follows.

    public void run() {
      hideCurrentPage();
      setCurrentPage(null);
    }

This way, when you call `Navigation.goTo` in a `redirect`, the `@PageHiding` method will not be called again as there will be no current page (and you will have already hidden it). But if `redirect` is never called, a `@PageHiding` method should still be called every time.